prometheus/registry.go

```diff
@@ -11,232 +11,253 @@
-	// for debugging.)
-	ProtoCompactTextTelemetryContentType = `application/vnd.google.protobuf; proto=io.prometheus.client.MetricFamily; encoding=compact-text`
-
-	// Constants for object pools.
-	numBufs           = 4
-	numMetricFamilies = 1000
-	numMetrics        = 10000
-
 	// Capacity for the channel to collect metrics and descriptors.
 	capMetricChan = 1000
 	capDescChan   = 10
+)
 
-	contentTypeHeader     = "Content-Type"
-	contentLengthHeader   = "Content-Length"
-	contentEncodingHeader = "Content-Encoding"
+// DefaultRegistry is the default registry implicitly used by a number of
+// convenience functions. It has a ProcessCollector and a GoCollector
+// pre-registered.
+var DefaultRegistry = NewRegistry()
 
-	acceptEncodingHeader = "Accept-Encoding"
-	acceptHeader         = "Accept"
-)
+func init() {
+	MustRegister(NewProcessCollector(os.Getpid(), ""))
+	MustRegister(NewGoCollector())
+}
 
-// Handler returns the HTTP handler for the global Prometheus registry. It is
-// already instrumented with InstrumentHandler (using "prometheus" as handler
-// name). Usually the handler is used to handle the "/metrics" endpoint.
-//
-// Please note the issues described in the doc comment of InstrumentHandler. You
-// might want to consider using UninstrumentedHandler instead.
-func Handler() http.Handler {
-	return InstrumentHandler("prometheus", defRegistry)
+// NewRegistry creates a new vanilla Registry without any Collectors
+// pre-registered.
+func NewRegistry() Registry {
+	return &registry{
+		collectorsByID:  map[uint64]Collector{},
+		descIDs:         map[uint64]struct{}{},
+		dimHashesByName: map[string]uint64{},
+	}
 }
 
-// UninstrumentedHandler works in the same way as Handler, but the returned HTTP
-// handler is not instrumented. This is useful if no instrumentation is desired
-// (for whatever reason) or if the instrumentation has to happen with a
-// different handler name (or with a different instrumentation approach
-// altogether). See the InstrumentHandler example.
-func UninstrumentedHandler() http.Handler {
-	return defRegistry
+// NewRegistryWithInjectionHook creates a registry with the provided hook to inject
+// MetricFamilies. The hook is a function that is called whenever metrics are
+// collected. The MetricFamily protobufs returned by the hook function are
+// merged with the metrics collected in the usual way.
+//
+// This is a way to directly inject MetricFamily protobufs managed and owned by
+// the caller. The caller has full responsibility. As no registration of the
+// injected metrics has happened, there was no check at registration-time. If
+// the injection results in inconsistent metrics, the Collect call will return
+// an error. Some problems may even go undetected, like invalid label names in
+// the injected protobufs.
+//
+// The hook function must be callable at any time and concurrently.
+func NewRegistryWithInjectionHook(hook func() []*dto.MetricFamily) Registry {
```

Comment:
That discussion was 3 years ago, lots has happened since then.

The Collect interface is already defined in terms of protos via Metric, converting doesn't seem like it should be a big deal.